File: food_recognition.py
import base64
from flask import request

import os

import vertexai
from vertexai.generative_models import GenerativeModel, Part, SafetySetting
from werkzeug.datastructures import FileStorage

from geminiTextRecognition import multiturn_generate_content


#This could be better but its just what the AI should be doing with the information like context.
textsi_1 = """We only want to find out what ingredients this person has by analyzing what is inside their fridge, 
pantry, or etc. Make it comma separated like: banana, apple, syrup... etc. Listing all the ingredients in their fridge"""


def generate(calories: str, protein: str, carbs: str, fat: str, people: str, image):
    os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = "useful-flame-441821-h0-5e6960a95210.json"

    vertexai.init(project="useful-flame-441821-h0", location="us-central1")
    model = GenerativeModel(
        "gemini-1.5-flash-002",
        system_instruction=[textsi_1]
    )

    pathName = "images/uploadedImage.png"
    with open(pathName, "rb") as imgFile:
        imageData = imgFile.read()


    #We encode the image
    encoded_image = base64.b64encode(imageData)

    image1 = Part.from_data(
        #This is the image type it is analyzing don't change it unless you change the image file type
        mime_type="image/png",
        #We Decode the image
        data=base64.b64decode(encoded_image),
    )

    #Needs fine tuning but it should be good after this
    responses = model.generate_content(
        [image1, """What is in this fridge?"""],
        generation_config=generation_config,
        safety_settings=safety_settings,
        stream=True,
    )

    #Gemini's image analysis is stored here
    fullResponse = ''
    for response in responses:
        fullResponse += response.text + ""

    #Removes image from our files
    if os.path.exists(pathName):
        os.remove(pathName)
    else:
        print("Unable to remove image")


    print("Full response: " + fullResponse)
    #we move into generating an actual response
    return multiturn_generate_content(fullResponse, calories, protein, carbs, fat, people)



generation_config = {
    "max_output_tokens": 8192,
    "temperature": 1,
    "top_p": 0.95,
}

safety_settings = [
    SafetySetting(
        category=SafetySetting.HarmCategory.HARM_CATEGORY_HATE_SPEECH,
        threshold=SafetySetting.HarmBlockThreshold.OFF
    ),
    SafetySetting(
        category=SafetySetting.HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT,
        threshold=SafetySetting.HarmBlockThreshold.OFF
    ),
    SafetySetting(
        category=SafetySetting.HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,
        threshold=SafetySetting.HarmBlockThreshold.OFF
    ),
    SafetySetting(
        category=SafetySetting.HarmCategory.HARM_CATEGORY_HARASSMENT,
        threshold=SafetySetting.HarmBlockThreshold.OFF
    ),
]

File: food_recognition.go
package main

import (
	"context"
	"encoding/base64"
	"fmt"
	"image"
	"image/png"
	"io"
	"os"

	aiplatform "cloud.google.com/go/aiplatform/apiv1"
	aiplatformpb "cloud.google.com/go/aiplatform/apiv1/aiplatformpb"
	"google.golang.org/api/option"
	structpb "google.golang.org/protobuf/types/known/structpb"
)

const projectID = "useful-flame-441821-h0"
const location = "us-east1"
const modelName = "gemini-1.5-flash-002"
const imagePath = "images/uploadedImage.png"

var textsi_food = `We only want to find out what ingredients this person has by analyzing what is inside their fridge, 
pantry, or etc. Make it comma separated like: banana, apple, syrup... etc. Listing all the ingredients in their fridge`

func mustOpen(imageFile string) *os.File {
	file, err := os.Open(imageFile)
	if err != nil {
		panic(err)
	}
	return file
}

func analyzeImg(imageFile string, reader io.Reader) (map[string]string, error) {
	//load google cloud  creds
	ctx := context.Background()
	client, err := aiplatform.NewPredictionClient(ctx, option.WithCredentialsFile("useful-flame-441821-h0-5e6960a95210.json"))
	if err != nil { //if there is an error
		return nil, fmt.Errorf("failed to create AI platform client: %v", err)
	}
	defer client.Close()

	// decode and save img to storage

	img, _, err := image.Decode(reader)
	if err != nil {
		return nil, fmt.Errorf("failed to decode image: %v", err)
	}

	outFile, err := os.Create(imagePath)
	if err != nil {
		return nil, fmt.Errorf("failed to create image file: %v", err)
	}
	defer outFile.Close()

	//then we encode the image as a png and we save it
	if err := png.Encode(outFile, img); err != nil {
		return nil, fmt.Errorf("failed to read image file: %v", err)
	}

	//next we read out the saved file
	imgData, err := os.ReadFile(imagePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read image file: %v", err)
	}

	//convert image to base64 encoding
	encodedImage := base64.StdEncoding.EncodeToString(imgData)

	// begin shenanigans
	// creates a request for vertex ai model
	params, err := structpb.NewStruct(map[string]interface{}{
		"temperature":       1.0,
		"max_output_tokens": 8192,
		"top_p":             0.95,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to create parameters struct: %v", err)
	}

	// now we create an instance struct
	instanceStruct, err := structpb.NewStruct(map[string]interface{}{
		"mime_type": "image.png",
		"data":      encodedImage,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to create instance struct: %v", err)
	}
	//this converts instancestruct to *structPB.value
	instanceValue, err := structpb.NewValue(instanceStruct)
	if err != nil {
		return nil, fmt.Errorf("failed to convert instance struct to value: %v", err)
	}
	//this converts params to *structpb.value
	paramsValue, err := structpb.NewValue(params)
	if err != nil {
		return nil, fmt.Errorf("failed to convert parameters struct to value: %v", err)
	}

	req := &aiplatformpb.PredictRequest{
		Endpoint: fmt.Sprintf("projects/%s/locations/%s/publishers/google/models/%s", projectID, location, modelName),
		Instances: []*structpb.Value{
			instanceValue,
		},
		Parameters: paramsValue,
	}

	// then we actually send the prediction request to Vertex AI
	resp, err := client.Predict(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("failed to predict: %v", err)
	}

	//collect the model's response
	fullResponse := ""
	for _, prediction := range resp.Predictions {
		fullResponse += prediction.String() + " "
	}

	//remove the saved image file to free up space
	if err := os.Remove(imagePath); err != nil {
		fmt.Println("Unable to remove image:", err)
	}

	//Return AI model's response as a json
	return map[string]string{"data": fullResponse}, nil
}

var generationConfig = map[string]interface{}{
	"max_output_tokens": 8192,
	"temperature":       1,
	"top_p":             0.95,
}

// Safety settings placeholder
var safetySettings = []string{
	"HARM_CATEGORY_HATE_SPEECH: OFF",
	"HARM_CATEGORY_DANGEROUS_CONTENT: OFF",
	"HARM_CATEGORY_SEXUALLY_EXPLICIT: OFF",
	"HARM_CATEGORY_HARASSMENT: OFF",
}
